File: library/src/main/java/me/shkschneider/skeleton/helper/VibratorHelper.java
package me.shkschneider.skeleton.helper;

import android.Manifest;
import android.annotation.TargetApi;
import android.os.VibrationEffect;
import android.os.Vibrator;
import android.support.annotation.RequiresPermission;

public class VibratorHelper {

    public static final long[] DEFAULT_DURATION = new long[] { 1000L, 1000L, 1000L, 1000L, 1000L };

    private static final int DO_REPEAT = 0;
    private static final int DO_NOT_REPEAT = -1;

    protected VibratorHelper() {
        // Empty
    }

    public static boolean has() {
        final Vibrator vibrator = SystemServices.vibrator();
        return (vibrator != null && vibrator.hasVibrator());
    }

    @SuppressWarnings("deprecation")
    @RequiresPermission(Manifest.permission.VIBRATE)
    public static void vibrate(final long[] durations, final boolean repeat) {
        if (AndroidHelper.api() >= AndroidHelper.API_26) {
            vibrate26(durations, repeat);
            return;
        }
        SystemServices.vibrator().vibrate(durations, (repeat ? DO_REPEAT : DO_NOT_REPEAT));
    }

    @TargetApi(AndroidHelper.API_26)
    @RequiresPermission(Manifest.permission.VIBRATE)
    public static void vibrate26(final long[] durations, final boolean repeat) {
        SystemServices.vibrator().vibrate(VibrationEffect.createWaveform(durations, (repeat ? DO_REPEAT : DO_NOT_REPEAT)));
    }

    @RequiresPermission(Manifest.permission.VIBRATE)
    public static void vibrate() {
        if (AndroidHelper.api() >= AndroidHelper.API_26) {
            vibrate26(DEFAULT_DURATION, false);
            return;
        }
        vibrate(DEFAULT_DURATION, false);
    }

    @TargetApi(AndroidHelper.API_26)
    @RequiresPermission(Manifest.permission.VIBRATE)
    public static void vibrate26() {
        vibrate26(DEFAULT_DURATION, false);
    }

}

File: library/src/main/kotlin/me/shkschneider/skeleton/helper/VibratorHelper.kt
package me.shkschneider.skeleton.helper

import android.Manifest
import android.annotation.SuppressLint
import android.os.Build
import android.os.VibrationEffect
import android.support.annotation.RequiresPermission

object VibratorHelper {

    private val DEFAULT_DURATION = longArrayOf(1000, 1000, 1000, 1000, 1000)
    private val DO_REPEAT = 0
    private val DO_NOT_REPEAT = -1

    fun has(): Boolean {
        val vibrator = SystemServices.vibrator()
        return vibrator != null && vibrator.hasVibrator()
    }

    @RequiresPermission(Manifest.permission.VIBRATE)
    fun vibrate(durations: LongArray, repeat: Boolean = false) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            SystemServices.vibrator()?.vibrate(VibrationEffect.createWaveform(durations, if (repeat) DO_REPEAT else DO_NOT_REPEAT))
        } else {
            SystemServices.vibrator()?.vibrate(durations, if (repeat) DO_REPEAT else DO_NOT_REPEAT)
        }
    }

    @RequiresPermission(Manifest.permission.VIBRATE)
    fun vibrate() {
        vibrate(DEFAULT_DURATION)
    }

}
